collect_app/src/main/java/org/odk/collect/android/fragments/dialogs/NumberPickerDialog.java

```diff
@@ -19,76 +19,68 @@
 import androidx.appcompat.app.AlertDialog;
 import android.app.Dialog;
 import android.content.Context;
-import android.content.DialogInterface;
 import android.os.Bundle;
 import androidx.fragment.app.DialogFragment;
+import androidx.lifecycle.ViewModelProvider;
+
 import android.view.LayoutInflater;
 import android.view.View;
 import android.widget.NumberPicker;
 
+import org.jetbrains.annotations.NotNull;
 import org.odk.collect.android.R;
+import org.odk.collect.android.widgets.viewmodels.RangePickerViewModel;
 
 public class NumberPickerDialog extends DialogFragment {
-
-    public static final String NUMBER_PICKER_DIALOG_TAG = "numberPickerDialogTag";
-    public static final String WIDGET_ID = "widgetId";
     public static final String DISPLAYED_VALUES = "displayedValues";
     public static final String PROGRESS = "progress";
 
-    public interface NumberPickerListener {
-        void onNumberPickerValueSelected(int widgetId, int value);
-    }
-
+    private RangePickerViewModel rangePickerViewModel;
     private NumberPickerListener listener;
 
-    public static NumberPickerDialog newInstance(int widgetId, String[] displayedValues, int progress) {
-        Bundle bundle = new Bundle();
-        bundle.putInt(WIDGET_ID, widgetId);
-        bundle.putSerializable(DISPLAYED_VALUES, displayedValues);
-        bundle.putInt(PROGRESS, progress);
-
-        NumberPickerDialog dialogFragment = new NumberPickerDialog();
-        dialogFragment.setArguments(bundle);
-        return dialogFragment;
-    }
-
     @Override
-    public void onAttach(Context context) {
+    public void onAttach(@NotNull Context context) {
         super.onAttach(context);
-
         try {
             listener = (NumberPickerListener) context;
         } catch (ClassCastException e) {
             throw new ClassCastException(context.toString() + " must implement OnHeadlineSelectedListener");
         }
+
+        rangePickerViewModel = new ViewModelProvider(this).get(RangePickerViewModel.class);
+        rangePickerViewModel.setDisplayedValuesForNumberPicker((String[]) getArguments().getSerializable(DISPLAYED_VALUES));
```

Comment:
Instead of passing this as an argument to the dialog and then to the view model, couldn't the data requester set the value on the view model?

Reply:
Yes correct. I was just confused that if I do so I will be initializing the viewmodel in `RangeWidgetDataRequester` class and then I will have to use the activity's (`FormEntryActivity`) as the owner of the viewmodel. Currently, the viewmodel lives until the dialog is shown and if the user opens up the `NumberPicker` again, a new viewmodel is instantiated and we don't have to worry about the viewmodel which was atttached to the `FormEntryActivity` previously.

I think if I use the activity as the owner, then I will be having the viewmodel as long as I am using the widgets of a particular form?

Reply:
Right I see. I think it'd be nice to get to a point where the ViewModel is the sole method of communication between the widget and the dialog but I it would also involve making the ViewModel work for multiple range widgets (like in a field list). For the moment then lets just go with the args being passed to the dialog and then then the dialog calling the FormEntryActivity via a listener.

Reply:
Oh okay yes I like he idea of having the viewmodel just the sole method of communication

Reply:
While I was working refactoring the Date and Time widgets, I came across the same problem when using the viewmodel. @grzesiek2010 guided and helped me in setting them up as I have done in this PR. However, he also liked the idea of excluding the activity from the conversation between viewmodel and widget. He told that later we will be using fragments to implement widgets, then it will be easy to use viewmodel without involving the activity.